imap_processing/codice/utils.py

```diff
@@ -57,3 +96,205 @@ class CoDICECompression(IntEnum):
+    return sci_lut_data
+
+
+def get_view_tab_info(json_data: dict, view_id: int, apid: int) -> dict:
+    """
+    Get the view table information for a specific view and APID.
+
+    Parameters
+    ----------
+    json_data : dict
+        The JSON data loaded from the SCI-LUT file.
+    view_id : int
+        The view ID from the packet.
+    apid : int
+        The APID from the packet.
+
+    Returns
+    -------
+    dict
+        The view table information containing details like sensor,
+        collapse_table, data_product, etc.
+    """
+    apid_hex = f"0x{apid:X}"
+    # This is how we get view information that will be used to get
+    # collapse pattern:
+    #  table_id -> view_tab -> (view_id, apid) -> sensor -> collapse_table
+    view_tab = json_data.get("view_tab").get(f"({view_id}, {apid_hex})")
+    return view_tab
+
+
+def get_collapse_pattern_shape(
+    json_data: dict, sensor_id: int, collapse_table_id: int
+) -> tuple[int, ...]:
+    """
+    Get the collapse pattern for a specific sensor id and collapse table id.
+
+    Parameters
+    ----------
+    json_data : dict
+        The JSON data loaded from the SCI-LUT file.
+    sensor_id : int
+        Sensor identifier (0 for LO, 1 for HI).
+    collapse_table_id : int
+        Collapse table id to look up in the SCI-LUT.
+
+    Returns
+    -------
+    tuple[int, ...]
+        The reduced shape describing the collapsed pattern. Examples:
+        ``(1,)`` for a fully collapsed 1-D pattern or ``(N, M)`` for a
+        reduced 2-D pattern.
+    """
+    if sensor_id == 0:
+        # LO sensor
+        collapse_tab = json_data.get("collapse_lo").get(f"{collapse_table_id}")
+    else:
+        # HI sensor
+        collapse_tab = json_data.get("collapse_hi").get(f"{collapse_table_id}")
+
+    # Analyze the collapse pattern matrix to determine its reduced shape.
+    # Steps:
+    # - Extract non-zero elements from the matrix.
+    # - Reshape to group unique non-zero rows and columns.
+    # - If all non-zero values are identical, return (1,) for a fully collapsed pattern.
+    # - Otherwise, compute the number of unique rows and columns to describe the
+    #   reduced shape.
+    collapse_matrix = np.array(collapse_tab["matrix"])
+    non_zero_data = np.where(collapse_matrix != 0)
+    non_zero_reformatted = collapse_matrix[non_zero_data].reshape(
+        np.unique(non_zero_data[0]).size, np.unique(non_zero_data[1]).size
+    )
+
+    if np.unique(non_zero_reformatted).size == 1:
+        # all non-zero values are identical means -> fully collapsed
+        return (1,)
+
+    # If not fully collapsed, find repeated patterns in rows and columns
+    # to reduce shape further.
+    unique_rows = np.unique(non_zero_reformatted, axis=0)
+    unique_columns = np.unique(non_zero_reformatted, axis=1)
+    # TODO: flip this dimension for final dataset to be below?
+    # (spin_sector, position or ssd id)
+    return (unique_rows.shape[0], unique_columns.shape[1])
```

Comment:
column is spin sector and row is az.
```suggestion
    return (unique_columns.shape[1], unique_rows.shape[0])
```